odoo/addons/base/models/ir_mail_server.py

```diff
@@ -556,6 +591,80 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
             raise MailDeliveryException(_("Mail Delivery Failed"), msg)
         return message_id
 
+    @api.model
+    def _find_mail_server(self, mail_from):
+        """Find the appropriate mail server for the given email address.
+
+        Returns: Record<ir.mail_server>, mail_from
+        - Mail server to use to send the email (None if we use the odoo-bin arguments)
+        - Email FROM to use to send the email (in some case, it might be impossible
+          to use the given email address directly if no mail server is configured for)
+        """
+        normalized_mail_from = email_normalize(mail_from)
+        mail_from_domain = email_domain_extract(normalized_mail_from)
+        notifications_email = email_normalize(self._get_default_from_address())
+        notifications_domain = email_domain_extract(notifications_email)
+        notifications_email_full = encapsulate_email(mail_from, notifications_email)
+
+        mail_servers = self.sudo().search([], order='sequence')
+
+        # 1. Try to find a mail server for the right mail from
+        mail_server = mail_servers.filtered(lambda m: email_normalize(m.from_filter) == normalized_mail_from)
+        if mail_server:
+            return mail_server[0], mail_from
+
+        mail_server = mail_servers.filtered(lambda m: email_domain_extract(m.from_filter) == mail_from_domain)
+        if mail_server:
+            return mail_server[0], mail_from
+
+        # 2. Try to find a mail server for <notifications@domain.com>
+        if notifications_email:
+            mail_server = mail_servers.filtered(lambda m: email_normalize(m.from_filter) == notifications_email)
+            if mail_server:
+                return mail_server[0], notifications_email_full
+
+            mail_server = mail_servers.filtered(lambda m: email_domain_extract(m.from_filter) == notifications_domain)
+            if mail_server:
+                return mail_server[0], notifications_email_full
+
+        # 3. Take the first mail server without "from_filter" because
```

Comment:
In case of "we spoof because no other choice", can't we fallback on behavior implemented in 12 (and reverted in this PR) ? Allowing a "backward compatibility". Or do we consider that this should be configured completely at server level ?

Reply:
IMO, it's better to not mix both solutions (otherwise the code will become less readable and the flow will be more complicated). It's the fault of the database administrator in that case, he should configure it properly...

Reply:
Ok for me.